src/commands/force/source/beta/pull.ts

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { FlagsConfig, flags } from '@salesforce/command';
+import { Duration } from '@salesforce/kit';
+import { Messages } from '@salesforce/core';
+import {
+  FileResponse,
+  SourceComponent,
+  ComponentSet,
+  RetrieveResult,
+  RequestStatus,
+  ComponentStatus,
+} from '@salesforce/source-deploy-retrieve';
+import { SourceTracking, throwIfInvalid, replaceRenamedCommands, ChangeResult } from '@salesforce/source-tracking';
+import { processConflicts } from '../../../../formatters/conflicts';
+import { SourceCommand } from '../../../../sourceCommand';
+import { PullResponse, PullResultFormatter } from '../../../../formatters/pullFormatter';
+Messages.importMessagesDirectory(__dirname);
+const messages: Messages = Messages.loadMessages('@salesforce/plugin-source', 'pull');
+
+export default class SourcePull extends SourceCommand {
+  public static description = messages.getMessage('description');
+  public static help = messages.getMessage('help');
+  protected static readonly flagsConfig: FlagsConfig = {
+    forceoverwrite: flags.boolean({
+      char: 'f',
+      description: messages.getMessage('flags.forceoverwrite'),
+    }),
+    // TODO: use shared flags from plugin-source
+    wait: flags.minutes({
+      char: 'w',
+      default: Duration.minutes(33),
+      min: Duration.minutes(0), // wait=0 means deploy is asynchronous
+      description: messages.getMessage('flags.waitLong'),
+    }),
+  };
+
+  protected static requiresUsername = true;
+  protected static requiresProject = true;
+  protected hidden = true;
+  protected readonly lifecycleEventNames = ['preretrieve', 'postretrieve'];
+  protected tracking: SourceTracking;
+  protected retrieveResult: RetrieveResult;
+  protected deleteFileResponses: FileResponse[];
+
+  public async run(): Promise<PullResponse[]> {
+    await this.preChecks();
+    await this.retrieve();
+    // do not parallelize delete and retrieve...we only get to delete IF retrieve was successful
+    await this.doDeletes(); // deletes includes its tracking file operations
+    await this.updateTrackingFilesWithRetrieve();
+    this.ux.stopSpinner();
+
+    return this.formatResult();
+  }
+
+  protected async preChecks(): Promise<void> {
+    // checks the source tracking file version and throws if they're toolbelt's old version
+    throwIfInvalid({
+      org: this.org,
+      projectPath: this.project.getPath(),
+      toValidate: 'plugin-source',
+      command: replaceRenamedCommands('force:source:pull'),
+    });
```

Comment:
It's a bit strange that STL does this but I understand the reasons why.  I think we should mark it in STL as deprecated from the start so we can eventually remove it after several releases.

Reply:
good idea.  stl has a Compatibility class for handling all transitional stuff (messaging, checking that you're not mixing files, etc).

Reply:
done here: https://github.com/forcedotcom/source-tracking/pull/20/commits/b68daec95b4fd9090e68eb28d15c96205a2b7dde